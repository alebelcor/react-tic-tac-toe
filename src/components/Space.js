import React from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import markSpace from '../redux/actions/markSpace';
import reset from '../redux/actions/reset';
import getIsGameOver from '../redux/selectors/getIsGameOver';

const Space = ({ hasBorderRight, hasBorderTop, handleClick, handleReset, isGameOver, markedBy, position }) => {
  const btnClass = classNames({
    'position-btn': true,
    'border-top': hasBorderTop,
    'border-right': hasBorderRight,
    'marked-1': markedBy === 1,
    'marked-2': markedBy === 2,
  });

  const onMarkSpace = () => {
    if (isGameOver) {
      return handleReset();
    }

    if (markedBy) {
      return undefined;
    }

    return handleClick();
  }

  return (
    <li className="position">
      <button
        className={btnClass}
        type="button"
        onClick={onMarkSpace}
      >
        <span className="sr-only">Position {position} of 9</span>
      </button>
    </li>
  );
}

const mapStateToProps = (state, ownProps) => {
  let markedBy;

  state.turns.some(turn => {
    if (turn.position === ownProps.position) {
      markedBy = turn.player;
      return true;
    }

    return false;
  });

  return {
    isGameOver: getIsGameOver(state),
    markedBy
  };
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  handleClick: () => dispatch(markSpace(ownProps.position)),
  handleReset: () => dispatch(reset())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Space);
