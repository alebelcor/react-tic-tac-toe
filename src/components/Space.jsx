/** @jsx jsx */
import { useMemo } from 'react';
import { jsx, css } from '@emotion/core';
import { useSelector, useDispatch } from 'react-redux';

import markSpace from '../redux/actions/markSpace';
import reset from '../redux/actions/reset';

import getIsGameOver from '../redux/selectors/getIsGameOver';
import getMarkedBy from '../redux/selectors/getMarkedBy';

import {
  positionStyle,
  positionButtonStyle,
  markedStyle,
} from '../styles/Space.css';

import {
  screenReaderOnlyStyle,
} from '../styles/utils.css';

const Space = ({ position }) => {
  const isGameOver = useSelector(getIsGameOver);

  const selectMarkedBy = useMemo(getMarkedBy, []);
  const markedBy = useSelector(
    state => selectMarkedBy(state, position)
  );

  const dispatch = useDispatch();
  const handleReset = () => {
    dispatch(reset());
  };

  const handleClick = (e) => {
    e.preventDefault();

    if (isGameOver) {
      handleReset();
      return;
    }

    if (markedBy > 0) {
      return;
    }

    dispatch(markSpace(position));
  };

  return (
    <li css={positionStyle}>
      <button
        css={css`
          ${positionButtonStyle};
          ${markedBy > 0 ? markedStyle : ``};
          ${!(position === 1 || position === 2 || position === 3) ? `border-top-width: 3px;` : ``};
          ${!(position === 3 || position === 6 || position === 9) ? `border-right-width: 3px;` : ``};
          ${isGameOver ? `cursor: initial;` : ``};
        `}
        type="button"
        aria-describedby={`position-${position}`}
        onClick={handleClick}
      >
        {markedBy > 0 && (
          <svg aria-hidden="true" fill={markedBy === 1 ? `#5bc0eb` : `#c3423f`} xmlns="http://www.w3.org/2000/svg" width="80" height="80" viewBox="0 0 24 24">
            {markedBy === 1 && <path d="M23 20.168l-8.185-8.187 8.185-8.174-2.832-2.807-8.182 8.179-8.176-8.179-2.81 2.81 8.186 8.196-8.186 8.184 2.81 2.81 8.203-8.192 8.18 8.192z" />}
            {markedBy === 2 && <path d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12z" />}
          </svg>
        )}

        <span id={`position-${position}`} css={screenReaderOnlyStyle}>
          Position {position} of 9

          {markedBy > 0 && (
            <span>, Selected by Player {markedBy}</span>
          )}
        </span>
      </button>
    </li>
  );
};

export default Space;
