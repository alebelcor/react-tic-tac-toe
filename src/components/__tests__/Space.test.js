import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';

import configureStore from '../../redux/configureStore';

import Space from '../Space';

describe('Space', () => {
  it('should render screen reader only text', () => {
    const position = 1;
    const store = configureStore({
      turns: [],
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <Space position={position} />
      </Provider>
    );

    const spaceEl = getByTestId(`space-${position}`);

    expect(spaceEl).toBeVisible();
    expect(spaceEl).toHaveTextContent(`Position ${position} of 9`);
  });

  it('should render a specific mark, when player 1 clicks it, and screen reader only text', () => {
    const position = 1;
    const player = 1;
    const store = configureStore({
      turns: [],
    });

    const { getByTestId, asFragment } = render(
      <Provider store={store}>
        <Space position={position} />
      </Provider>
    );

    fireEvent.click(getByTestId(`space-${position}`));

    expect(getByTestId(`space-${position}`)).toHaveTextContent(`Position ${position} of 9, Selected by Player ${player}`);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render a specific mark, when player 2 clicks it, and screen reader only text', () => {
    const position = 2;
    const player = 2;
    const store = configureStore({
      turns: [{ player: 1, position: 1, },],
    });

    const { getByTestId, asFragment } = render(
      <Provider store={store}>
        <Space position={position} />
      </Provider>
    );

    fireEvent.click(getByTestId(`space-${position}`));

    expect(getByTestId(`space-${position}`)).toHaveTextContent(`Position ${position} of 9, Selected by Player ${player}`);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should do nothing, when clicked, if already selected', () => {
    const player = 1;
    const position = 1;
    const store = configureStore({
      turns: [{ player, position, },],
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <Space position={position} />
      </Provider>
    );

    fireEvent.click(getByTestId(`space-${position}`));

    const spaceEl = getByTestId(`space-${position}`);
    expect(spaceEl).toBeVisible();
    expect(spaceEl).toHaveTextContent(`Position ${position} of 9, Selected by Player ${player}`);
  });

  it('should reset when game is over', () => {
    const position = 1;
    const store = configureStore({
      turns: [{}, {}, {}, {}, {}, {}, {}, {}, {},],
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <Space position={position} />
      </Provider>
    );

    fireEvent.click(getByTestId(`space-${position}`));

    const spaceEl = getByTestId(`space-${position}`);
    expect(spaceEl).toBeVisible();
    expect(spaceEl).toHaveTextContent(`Position ${position} of 9`);
  });
});
