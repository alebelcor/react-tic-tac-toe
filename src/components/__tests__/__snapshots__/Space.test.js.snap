// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Space should render a specific mark, when player 1 clicks it, and screen reader only text 1`] = `
<DocumentFragment>
  <li
    class="css-1v8pj0r-Space"
  >
    <button
      aria-describedby="position-1"
      class="css-xy2ha3-Space"
      data-testid="space-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="#5bc0eb"
        height="80"
        viewBox="0 0 24 24"
        width="80"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M23 20.168l-8.185-8.187 8.185-8.174-2.832-2.807-8.182 8.179-8.176-8.179-2.81 2.81 8.186 8.196-8.186 8.184 2.81 2.81 8.203-8.192 8.18 8.192z"
        />
      </svg>
      <span
        class="css-1qsia7v-Space"
        id="position-1"
      >
        Position 1 of 9
        <span>
          , Selected by Player 1
        </span>
      </span>
    </button>
  </li>
</DocumentFragment>
`;

exports[`Space should render a specific mark, when player 2 clicks it, and screen reader only text 1`] = `
<DocumentFragment>
  <li
    class="css-1v8pj0r-Space"
  >
    <button
      aria-describedby="position-2"
      class="css-xy2ha3-Space"
      data-testid="space-2"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="#c3423f"
        height="80"
        viewBox="0 0 24 24"
        width="80"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 2c5.514 0 10 4.486 10 10s-4.486 10-10 10-10-4.486-10-10 4.486-10 10-10zm0-2c-6.627 0-12 5.373-12 12s5.373 12 12 12 12-5.373 12-12-5.373-12-12-12z"
        />
      </svg>
      <span
        class="css-1qsia7v-Space"
        id="position-2"
      >
        Position 2 of 9
        <span>
          , Selected by Player 2
        </span>
      </span>
    </button>
  </li>
</DocumentFragment>
`;
